velox/functions/sparksql/tests/SparkPartitionIdTest.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/functions/sparksql/tests/SparkFunctionBaseTest.h"
+
+namespace facebook::velox::functions::sparksql::test {
+namespace {
+
+class SparkPartitionIdTest : public SparkFunctionBaseTest {
+ protected:
+  void setSparkPartitionId(int32_t partitionId) {
+    queryCtx_->testingOverrideConfigUnsafe(
+        {{core::QueryConfig::kSparkPartitionId, std::to_string(partitionId)}});
+  }
+
+  void testSparkPartitionId(int32_t partitionId, int32_t vectorSize) {
+    setSparkPartitionId(partitionId);
+    auto result =
+        evaluate("spark_partition_id()", makeRowVector(ROW({}), vectorSize));
+    ASSERT_TRUE(result->isConstantEncoding());
+    velox::test::assertEqualVectors(
+        makeConstant(partitionId, vectorSize), result);
+  }
+};
+
+TEST_F(SparkPartitionIdTest, basic) {
+  testSparkPartitionId(0, 1);
+  testSparkPartitionId(100, 1);
+  testSparkPartitionId(0, 100);
+  testSparkPartitionId(100, 100);
+}
+
+TEST_F(SparkPartitionIdTest, error) {
+  auto rowVector = makeRowVector(ROW({}), 1);
+
+  queryCtx_->testingOverrideConfigUnsafe({{}});
+  VELOX_ASSERT_THROW(
+      evaluate("spark_partition_id()", rowVector),
+      "Spark partition id is not set");
+
+  setSparkPartitionId(-1);
+  VELOX_ASSERT_THROW(
+      evaluate("spark_partition_id()", rowVector),
```

Comment:
would you add a test for try(spark_partition_id()) to make sure these errors are supressed?

Reply:
I tested that `try` can not supress the error, is any thing I missed? below is the stack trace, thanks.
```
E0307 06:50:42.610155 1045883 Exceptions.h:69] Line: /var/git/velox/./velox/core/QueryConfig.h:674, Function:sparkPartitionId, Expression: id.has_value() Spark partition id is not set., Source: RUNTIME, ErrorCode: INVALID_STATE
E0307 06:50:42.610558 1045883 Exceptions.h:69] Line: /var/git/velox/./velox/core/QueryConfig.h:674, Function:sparkPartitionId, Expression: id.has_value() Spark partition id is not set., Source: RUNTIME, ErrorCode: INVALID_STATE
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Spark partition id is not set.
Retriable: False
Expression: id.has_value()
Function: sparkPartitionId
File: /var/git/velox/./velox/core/QueryConfig.h
Line: 674
Stack trace:
# 0  facebook::velox::VeloxException::VeloxException(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)
# 1  void facebook::velox::detail::veloxCheckFail<facebook::velox::VeloxRuntimeError, char const*>(facebook::velox::detail::VeloxCheckFailArgs const&, char const*)
# 2  facebook::velox::exec::SimpleFunctionAdapterFactoryImpl<facebook::velox::core::UDFHolder<facebook::velox::functions::sparksql::SparkPartitionIdFunction<facebook::velox::exec::VectorExec>, facebook::velox::exec::VectorExec, int, facebook::velox::ConstantChecker<>> >::createVectorFunction(std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > > const&, facebook::velox::core::QueryConfig const&) const
# 3  facebook::velox::exec::(anonymous namespace)::compileRewrittenExpression(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, facebook::velox::exec::(anonymous namespace)::Scope*, facebook::velox::core::QueryConfig const&, facebook::velox::memory::MemoryPool*, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool)
# 4  facebook::velox::exec::(anonymous namespace)::compileExpression(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, facebook::velox::exec::(anonymous namespace)::Scope*, facebook::velox::core::QueryConfig const&, facebook::velox::memory::MemoryPool*, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool)
# 5  facebook::velox::exec::(anonymous namespace)::compileRewrittenExpression(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, facebook::velox::exec::(anonymous namespace)::Scope*, facebook::velox::core::QueryConfig const&, facebook::velox::memory::MemoryPool*, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool)
# 6  facebook::velox::exec::(anonymous namespace)::compileExpression(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, facebook::velox::exec::(anonymous namespace)::Scope*, facebook::velox::core::QueryConfig const&, facebook::velox::memory::MemoryPool*, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool)
# 7  facebook::velox::exec::compileExpressions(std::vector<std::shared_ptr<facebook::velox::core::ITypedExpr const>, std::allocator<std::shared_ptr<facebook::velox::core::ITypedExpr const> > > const&, facebook::velox::core::ExecCtx*, facebook::velox::exec::ExprSet*, bool)
# 8  facebook::velox::exec::ExprSet::ExprSet(std::vector<std::shared_ptr<facebook::velox::core::ITypedExpr const>, std::allocator<std::shared_ptr<facebook::velox::core::ITypedExpr const> > > const&, facebook::velox::core::ExecCtx*, bool)
# 9  facebook::velox::functions::test::FunctionBaseTest::evaluate(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
# 10 facebook::velox::functions::sparksql::test::(anonymous namespace)::SparkPartitionIdTest_error_Test::TestBody() [clone .cold]
# 11 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
# 12 testing::Test::Run()
# 13 testing::TestInfo::Run()
# 14 testing::TestSuite::Run()
# 15 testing::internal::UnitTestImpl::RunAllTests()
# 16 testing::UnitTest::Run()
# 17 main
# 18 0x0000000000029d8f
# 19 __libc_start_main
# 20 _start
" thrown in the test body.
```

Reply:
I see. This is probably because you throw runtime not user error. VELOX_FAIL vs. VELOX_USER_FAIL

Reply:
Maybe that's expected since this is not really a user error.